README.md
Bpm’online
==========

[![License](https://poser.pugx.org/ossbrownie/bpm-online/license)](https://packagist.org/packages/ossbrownie/bpm-online)
[![Latest Stable Version](https://poser.pugx.org/ossbrownie/bpm-online/v/stable)](https://packagist.org/packages/ossbrownie/bpm-online)
[![Latest Unstable Version](https://poser.pugx.org/ossbrownie/bpm-online/v/unstable)](https://packagist.org/packages/ossbrownie/bpm-online)
[![Total Downloads](https://poser.pugx.org/ossbrownie/bpm-online/downloads)](https://packagist.org/packages/ossbrownie/bpm-online)

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/ossbrownie/bpm-online/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/ossbrownie/bpm-online/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/ossbrownie/bpm-online/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/ossbrownie/bpm-online/?branch=master)
[![Code Intelligence Status](https://scrutinizer-ci.com/g/ossbrownie/bpm-online/badges/code-intelligence.svg?b=master)](https://scrutinizer-ci.com/code-intelligence)
[![Build Status](https://travis-ci.org/ossbrownie/bpm-online.svg?branch=master)](https://travis-ci.org/ossbrownie/bpm-online)

[![PHP Version](https://img.shields.io/badge/PHP-%3E%3D7.0-brightgreen.svg)](https://php.net/)

API access to a single marketing management platform 'CRM-system bpm’online' >= 7.10

## Requirements
- **PHP** = >= 7.0
- **EXT-CURL** = *
- **EXT-JSON** = *
- **ossbrownie/http-client** = 0.0.7
- **ossbrownie/util** = 0.0.5

For more information about the extensions see:

**curl** - [https://www.php.net/curl](https://www.php.net/curl)

**json** - [https://www.php.net/json](https://www.php.net/json)


## Installation
Add a line to your "require" section in your composer configuration:
```json
{
    "require": {
        "ossbrownie/bpm-online": "0.0.8"
    }
}
```


## Documentation
- [DataService Contract InsertQuery.](https://github.com/ossbrownie/bpm-online/wiki/Data-Contract-InsertQuery) 
- [DataService Contract SelectQuery.](https://github.com/ossbrownie/bpm-online/wiki/Data-Contract-SelectQuery) 
- [DataService Contract UpdateQuery.](https://github.com/ossbrownie/bpm-online/wiki/Data-Contract-UpdateQuery) 
- [DataService Contract DeleteQuery.](https://github.com/ossbrownie/bpm-online/wiki/Data-Contract-DeleteQuery) 
- [DataService Contract BatchQuery.](https://github.com/ossbrownie/bpm-online/wiki/Data-Contract-BatchQuery) 
- [DataService DateTime.](https://github.com/ossbrownie/bpm-online/wiki/DateTime) 
- [DataService Filters.](https://github.com/ossbrownie/bpm-online/wiki/Filters)


## Tests
To run the test suite, you need install the dependencies via composer, then run PHPUnit.
```bash
$> composer.phar install
$> ./vendor/bin/phpunit --bootstrap ./tests/bootstrap.php ./tests
```


## License
Bpm’online is licensed under the [MIT License](https://opensource.org/licenses/MIT)


## Contact
Problems, comments, and suggestions all welcome: [oss.brownie@gmail.com](mailto:oss.brownie@gmail.com)


CHANGELOG.md
# Change Log

## dev-master (2019-06-12)

## 0.0.8 (2019-06-12)

- Updated CHANGELOG.md and README.md.
- Fixed key case in differents ResponseStatus.
- Changed unit tests.

## 0.0.7 (2019-03-27)

- Customized work with the filter IN.
- Added class Brownie\BpmOnline\DataService\Column\ColumnExpressionCollection and unitTest for it.

## 0.0.6 (2019-03-25)

- Added a batch request method to get the number of added contracts.

## 0.0.5 (2019-03-22)

- Updated composer.json
- Added up to 3 request attempts if http code 200,500 is not received.

## 0.0.4 (2019-03-20)

- The DateTime class has been added to form time in queries.
- Added class Brownie\BpmOnline\Util\DateTime
- Added unitTest for class Brownie\BpmOnline\Util\DateTime
- Fixed unit tests.

## 0.0.3 (2019-03-16)

- Added group filters.
- Added expressions for filters.
- Added batch query.
- Updated unitTest.

## 0.0.2 (2019-03-14)

- Updated unitTest and fixed scrutinizer issues.

## 0.0.1 (2019-03-13)

- Refactoring.
- Added class Brownie\BpmOnline\DataService\Response\DeleteContract
- Added class Brownie\BpmOnline\DataService\Response\InsertContract
- Added class Brownie\BpmOnline\DataService\Response\SelectContract
- Added class Brownie\BpmOnline\DataService\Response\UpdateContract
- Added class Brownie\BpmOnline\DataService\Response
- Added unitTest for class Brownie\BpmOnline\DataService\Response\DeleteContract
- Added unitTest for class Brownie\BpmOnline\DataService\Response\InsertContract
- Added unitTest for class Brownie\BpmOnline\DataService\Response\SelectContract
- Added unitTest for class Brownie\BpmOnline\DataService\Response\UpdateContract
- Added unitTest for class Brownie\BpmOnline\DataService\Response

## dev-master (2019-03-12)

- Refactoring.
- Updated composer.json, .scrutinizer.yml, .travis.yml
- Update annotations.

## dev-master (2019-03-11)

- Refactoring.
- Added unitTest for class Brownie\BpmOnline\DataService\Column\ColumnExpression
- Added unitTest for class Brownie\BpmOnline\DataService\Column\ColumnFilter
- Added unitTest for class Brownie\BpmOnline\DataService\Column\Expression\ColumnPath
- Added unitTest for class Brownie\BpmOnline\DataService\Column\Expression\ExpressionType
- Added unitTest for class Brownie\BpmOnline\DataService\Column\Expression\Parameter
- Added unitTest for class Brownie\BpmOnline\DataService\Column\Expression
- Added unitTest for class Brownie\BpmOnline\DataService\Contract\DeleteContract
- Added unitTest for class Brownie\BpmOnline\DataService\Contract\InsertContract
- Added unitTest for class Brownie\BpmOnline\DataService\Contract\SelectContract
- Added unitTest for class Brownie\BpmOnline\DataService\Contract\UpdateContract
- Added unitTest for class Brownie\BpmOnline\DataService\Contract
- Updated composer.json, .gitignore

## dev-master (2019-03-08)

Features:
- Initinal commit.
- Added authentication.
- Added minimal DataContract InsertQuery.
- Added minimal DataContract SelectQuery.
- Added minimal DataContract UpdateQuery.
- Added minimal DataContract DeleteQuery.
- Added minimal data Filters.
- Added unitTest for class Brownie\BpmOnline\Config.
- Added unitTest for class Brownie\BpmOnline\Exception\AuthenticationException.
- Added unitTest for class Brownie\BpmOnline\Exception\HttpCodeException.
- Added unitTest for class Brownie\BpmOnline\Exception\JsonException.
- Added unitTest for class Brownie\BpmOnline\Exception\ValidateException.


src/Brownie/BpmOnline/BpmOnline.php
<?php
/**
 * @category    Brownie/BpmOnline
 * @author      Brownie <oss.brownie@gmail.com>
 * @license     https://opensource.org/licenses/MIT
 */

namespace Brownie\BpmOnline;

use Brownie\BpmOnline\Exception\AuthenticationException;
use Brownie\BpmOnline\Exception\HttpCodeException;
use Brownie\HttpClient\Header\Header;
use Brownie\HttpClient\HTTPClient;
use Brownie\BpmOnline\DataService\Contract;
use Brownie\HttpClient\Request;
use Brownie\Util\StorageArray;
use Brownie\HttpClient\Exception\ClientException;
use Brownie\HttpClient\Exception\ValidateException;
use Brownie\BpmOnline\DataService\Response;

/**
 * API access to a single marketing management platform 'CRM-system bpm’online'
 *
 * @method HTTPClient   getHttpClient()             HTTP client.
 * @method Config       getConfig()                 Bpm'online config.
 * @method array        getAuthentication()         Returns authentication data.
 * @method BpmOnline    setAuthentication($data)    Sets authentication data.
 * @method int          getConfigurationNumber()    Returns configuration number.
 * @method string       getDataFormat()             Returns data format.
 */
class BpmOnline extends StorageArray
{

    /**
     * List of supported fields.
     *
     * @var array
     */
    protected $fields = [
        'httpClient' => null,
        'config' => null,
        'configurationNumber' => 0,
        'dataFormat' => 'json',
        'authentication' => null,
    ];

    /**
     * Sets the input values.
     *
     * @param HTTPClient    $httpClient     HTTP client.
     * @param Config        $config         Bpm'online config.
     */
    public function __construct(HTTPClient $httpClient, Config $config)
    {
        parent::__construct([
            'httpClient' => $httpClient,
            'config' => $config,
        ]);
    }

    /**
     * Creates a request to DataService.
     *
     * @param string    $url        Request URL.
     * @param string    $rawBody    Raw body.
     *
     * @return Request
     */
    private function createRequest($url, $rawBody)
    {
        return $this
            ->getHttpClient()
            ->createRequest()
            ->setMethod(Request::HTTP_METHOD_POST)
            ->setUrl($url)
            ->setBodyFormat(Request::FORMAT_APPLICATION_JSON)
            ->addHeader(new Header([
                'name' => 'Content-Type',
                'value' => Request::FORMAT_APPLICATION_JSON,
            ]))
            ->setBody($rawBody)
            ->setTimeOut($this->getConfig()->getApiConnectTimeOut());
    }

    /**
     * Authenticates in the DataService bpm'online.
     *
     * @throws AuthenticationException
     * @throws HttpCodeException
     * @throws ClientException
     * @throws ValidateException
     */
    private function authentication()
    {
        $request = $this
            ->createRequest(
                $this->getConfig()->getApiUrlScheme() . '://' .
                $this->getConfig()->getUserDomain() . '.' .
                $this->getConfig()->getApiDomain() .
                '/ServiceModel/AuthService.svc/Login',
                json_encode([
                    'UserName' => $this->getConfig()->getUserName(),
                    'UserPassword' => $this->getConfig()->getUserPassword(),
                ])
            );

        $response = $this->getHttpClient()->request($request);

        if (200 != $response->getHttpCode()) {
            throw new HttpCodeException('Invalid response code: ' . $response->getHttpCode());
        }

        $jsonResponse = json_decode($response->getBody(), true);

        if ((JSON_ERROR_NONE != json_last_error()) || (0 != $jsonResponse['Code']))
        {
            throw new AuthenticationException('Authentication failed.');
        }

        $this->setAuthentication([
            '.aspxauth' => $response->getHttpCookieList()->get('.aspxauth'),
            'bpmcsrf' => $response->getHttpCookieList()->get('bpmcsrf'),
        ]);
    }

    /**
     * Returns the response from the execution of the contract.
     *
     * @param Contract  $contract    DataService contract.
     * @param int       $attempts    Number of request retries.
     *
     * @return Response
     *
     * @throws AuthenticationException
     * @throws ClientException
     * @throws Exception\ValidateException
     * @throws HttpCodeException
     * @throws ValidateException
     */
    public function getResponse(Contract $contract, $attempts = 3)
    {
        $contract->validate();
        if (empty($this->getAuthentication())) {
            $this->authentication();
        }

        $request = $this
            ->createRequest(
                $this->getConfig()->getApiUrlScheme() . '://' .
                $this->getConfig()->getUserDomain() . '.' . $this->getConfig()->getApiDomain() .
                '/' . $this->getConfigurationNumber() . '/dataservice/' . $this->getDataFormat() .
                '/reply/' . $contract->getContractType(),
                json_encode($contract->toArray())
            )
            ->addCookie($this->getAuthentication()['.aspxauth'])
            ->addCookie($this->getAuthentication()['bpmcsrf'])
            ->addHeader(
                $this
                    ->getHttpClient()
                    ->createHeader('Authorization', 'Cookie')
            )
            ->addHeader(
                $this
                    ->getHttpClient()
                    ->createHeader('BPMCSRF', $this->getAuthentication()['bpmcsrf']->getValue())
            );

        $response = $this->getHttpClient()->request($request, $attempts, [200, 500]);
        $contractResponse = $contract->getResponse($response->getBody());

        if (200 != $response->getHttpCode()) {
            throw new HttpCodeException(
                'Invalid response code: ' . $response->getHttpCode() . ', ' .
                $contractResponse->getErrorMessage()
            );
        }

        return $contractResponse;
    }
}
